File: Ruler/CSharp/Ruler/MyModel.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Input;
using devDept.Eyeshot;
using devDept.Graphics;
using devDept.Geometry;
using System.Drawing;

namespace WpfApplication1
{
    public class MyModel : devDept.Eyeshot.Model
    {
        Point2D lowerLeftWorldCoord, upperRightWorldCoord;
        Point2D originScreen;

        public enum rulerPlaneType
        {
            ZX, YZ, XY
        }

        public rulerPlaneType RulerPlaneMode;
        Plane rulerPlane;

        protected override void DrawViewport(DrawSceneParams myParams)
        {
            base.DrawViewport(myParams);

            switch (RulerPlaneMode)
            {
                case rulerPlaneType.XY:
                    rulerPlane = Plane.XY;
                    break;
                case rulerPlaneType.ZX:
                    rulerPlane = Plane.ZX;
                    break;
                case rulerPlaneType.YZ:
                    rulerPlane = Plane.YZ;
                    break;
            }

            // Get the world coordinates of the corners and the screen position of the origin for reference

            Point3D ptUpperRight, ptLowerLeft;            
            ScreenToPlane(new System.Drawing.Point((int)Size.Width, rulerSize), rulerPlane, out ptUpperRight);
            ScreenToPlane(new System.Drawing.Point(rulerSize, (int)Size.Height), rulerPlane, out ptLowerLeft);

            originScreen = WorldToScreen(0, 0, 0);

            switch (RulerPlaneMode)
            {
                case rulerPlaneType.XY:
                    upperRightWorldCoord = new Point2D(ptUpperRight.X, ptUpperRight.Y);
                    lowerLeftWorldCoord = new Point2D(ptLowerLeft.X, ptLowerLeft.Y);
                    break;
                case rulerPlaneType.ZX:
                    upperRightWorldCoord = new Point2D(ptUpperRight.X, ptUpperRight.Z);
                    lowerLeftWorldCoord = new Point2D(ptLowerLeft.X, ptLowerLeft.Z);
                    break;
                case rulerPlaneType.YZ:
                    upperRightWorldCoord = new Point2D(ptUpperRight.Y, ptUpperRight.Z);
                    lowerLeftWorldCoord = new Point2D(ptLowerLeft.Y, ptLowerLeft.Z);
                    break;
            }

        }

        int rulerSize = 35;

        protected override void DrawOverlay(DrawSceneParams myParams)
        {
            double Height = Size.Height;
            double Width = Size.Width;

            renderContext.SetState(depthStencilStateType.DepthTestOff);
            renderContext.SetState(blendStateType.Blend);

            // Draw the transparent ruler
            renderContext.SetColorWireframe(Color.FromArgb((int)(0.4 * 255), 255, 255, 255));
            renderContext.SetState(rasterizerStateType.CCW_PolygonFill_NoCullFace_NoPolygonOffset);

            // Vertical Ruler
            renderContext.DrawQuad(new RectangleF(0, 0, rulerSize, (float) (Height - rulerSize)));

            // Horizontal Ruler
            renderContext.DrawQuad(new RectangleF(rulerSize, (float)(Height - rulerSize), (float)(Width - rulerSize), rulerSize));                       

            renderContext.SetState(blendStateType.NoBlend);

            // choose a string format with 2 decimal numbers
            string formatString = "{0:0.##}";

            double worldToScreen = (Height - rulerSize) / (upperRightWorldCoord.Y - lowerLeftWorldCoord.Y);

            double stepWorldX = 5, stepWorldY = 5;

            double worldHeight = upperRightWorldCoord.Y - lowerLeftWorldCoord.Y;
            double nlinesH = (worldHeight / stepWorldY);

            double worldWidth = upperRightWorldCoord.X - lowerLeftWorldCoord.X;
            double nlinesW = (worldWidth / stepWorldX);

            RefineStep(nlinesH, worldHeight, ref stepWorldY);
            RefineStep(nlinesW, worldWidth, ref stepWorldX);

            double stepWorld = Math.Min(stepWorldX, stepWorldY);

            double stepScreen = stepWorld * worldToScreen;

            ///////////////////////////
            // Vertical ruler
            ///////////////////////////

            // First line Y world coordinate
            double startYWorld = stepWorld * Math.Floor(lowerLeftWorldCoord.Y / stepWorld);

            Point2D firstLineScreenPositionY = new Point2D(0, originScreen.Y + startYWorld * worldToScreen);
            double currentScreenY = firstLineScreenPositionY.Y;
            double shorterLineXPos = (firstLineScreenPositionY.X + rulerSize) / 2;

            // draw a longer line each 5 lines. The origin must be a longer line.
            int countShortLinesY = (int)(Math.Round((currentScreenY - originScreen.Y) / stepScreen)) % 5;

            // Draw the ruler lines
            renderContext.SetLineSize(1);

            double left;

            Font myFont = UtilityEx.GetFont(FontFamily, FontStyle, FontWeight, FontSize);

            for (double y = startYWorld; y < upperRightWorldCoord.Y; y += stepWorld, currentScreenY += stepScreen)
            {
                if (countShortLinesY % 5 == 0)
                    left = firstLineScreenPositionY.X;
                else
                    left = shorterLineXPos; ;

                renderContext.SetColorWireframe(Color.Black);

                renderContext.DrawLine(new Point2D(left, currentScreenY), new Point2D(rulerSize, currentScreenY));

                DrawText((int)firstLineScreenPositionY.X, (int)currentScreenY, string.Format(formatString, y), myFont, Color.Black, ContentAlignment.BottomLeft);

                countShortLinesY++;
            }


            ///////////////////////////
            // Horizontal ruler
            ///////////////////////////

            // First line X world coordinate
            double startXWorld = stepWorld * Math.Ceiling(lowerLeftWorldCoord.X / stepWorld);

            Point2D firstLineScreenPositionX = new Point2D(originScreen.X + startXWorld * worldToScreen, 0);
            double currentScreenX = firstLineScreenPositionX.X;
            double shorterLineYPos = (firstLineScreenPositionX.Y + rulerSize) / 2;

            int countShortLinesX = (int)(Math.Round((currentScreenX - originScreen.X) / stepScreen)) % 5;

            double top;
            for (double x = startXWorld; x < upperRightWorldCoord.X; x += stepWorld, currentScreenX += stepScreen)
            {
                if (countShortLinesX % 5 == 0)
                    top = firstLineScreenPositionX.Y;
                else
                    top = shorterLineYPos;

                renderContext.SetColorWireframe(Color.Black);

                renderContext.DrawLine(new Point2D(currentScreenX, Height - top), new Point2D(currentScreenX, Height - rulerSize));

                DrawText((int)currentScreenX, (int)(Height - rulerSize - firstLineScreenPositionX.Y), string.Format(formatString, x), myFont, Color.Black, ContentAlignment.BottomLeft);

                countShortLinesX++;
            }

            // Draw a red line in correspondence with the mouse position
            
            renderContext.SetColorWireframe(Color.Red);
            
            if (mousePos.Y > rulerSize)
            {
                renderContext.DrawLine(new Point3D(0, Height - mousePos.Y, 0), new Point3D(rulerSize, Height - mousePos.Y, 0));
            }

            if (mousePos.X > rulerSize)
            {
                renderContext.DrawLine(new Point3D(mousePos.X, Height, 0), new Point3D(mousePos.X, Height - rulerSize, 0));
            }

            // Draw the logo image at the bottom right corner            
            Bitmap logo = new Bitmap("../../../../../../dataset/Assets/Pictures/Logo.png");
            DrawImage(Size.Width - logo.Width - 20, 20, logo);

            // call the base function
            base.DrawOverlay(myParams);
        }

        private static void RefineStep(double nlines, double worldLength, ref double stepWorld)
        {
            // Refine the step if there are too many ruler lines, or too few
            if (nlines > 20)
            {
                do
                {
                    stepWorld *= 2;
                    nlines = (worldLength/stepWorld);
                } while (nlines > 20);
            }
            else if (nlines < 10)
            {
                do
                {
                    stepWorld /= 2;
                    nlines = (worldLength/stepWorld);
                } while (nlines < 10 && nlines > 0);
            }            
        }

        private System.Drawing.Point mousePos;

        protected override void OnMouseMove(MouseEventArgs e)
        {            
            base.OnMouseMove(e);

            mousePos = RenderContextUtility.ConvertPoint(GetMousePosition(e));

            PaintBackBuffer();
            SwapBuffers();
        }
    }

}

File: Ruler/Visual Basic/Ruler/MyModel.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Input
Imports devDept.Eyeshot
Imports devDept.Graphics
Imports devDept.Geometry
Imports System.Drawing

Public Class MyModel
    Inherits devDept.Eyeshot.Model
    Private lowerLeftWorldCoord As Point2D, upperRightWorldCoord As Point2D
    Private originScreen As Point2D

    Public Enum rulerPlaneType
        ZX
        YZ
        XY
    End Enum

    Public RulerPlaneMode As rulerPlaneType
    Private rulerPlane As Plane

    Protected Overrides Sub DrawViewport(myParams As DrawSceneParams)
        MyBase.DrawViewport(myParams)

        Select Case RulerPlaneMode
            Case rulerPlaneType.XY
                rulerPlane = Plane.XY
                Exit Select
            Case rulerPlaneType.ZX
                rulerPlane = Plane.ZX
                Exit Select
            Case rulerPlaneType.YZ
                rulerPlane = Plane.YZ
                Exit Select
        End Select

        ' Get the world coordinates of the corners and the screen position of the origin for reference

        Dim ptUpperRight As Point3D, ptLowerLeft As Point3D
        ScreenToPlane(New System.Drawing.Point(CInt(Size.Width), rulerSize), rulerPlane, ptUpperRight)
        ScreenToPlane(New System.Drawing.Point(rulerSize, CInt(Size.Height)), rulerPlane, ptLowerLeft)

        originScreen = WorldToScreen(0, 0, 0)

        Select Case RulerPlaneMode
            Case rulerPlaneType.XY
                upperRightWorldCoord = New Point2D(ptUpperRight.X, ptUpperRight.Y)
                lowerLeftWorldCoord = New Point2D(ptLowerLeft.X, ptLowerLeft.Y)
                Exit Select
            Case rulerPlaneType.ZX
                upperRightWorldCoord = New Point2D(ptUpperRight.X, ptUpperRight.Z)
                lowerLeftWorldCoord = New Point2D(ptLowerLeft.X, ptLowerLeft.Z)
                Exit Select
            Case rulerPlaneType.YZ
                upperRightWorldCoord = New Point2D(ptUpperRight.Y, ptUpperRight.Z)
                lowerLeftWorldCoord = New Point2D(ptLowerLeft.Y, ptLowerLeft.Z)
                Exit Select
        End Select

    End Sub

    Private rulerSize As Integer = 35

    Protected Overrides Sub DrawOverlay(myParams As DrawSceneParams)
        Dim Height As Double = Size.Height
        Dim Width As Double = Size.Width

        renderContext.SetState(depthStencilStateType.DepthTestOff)
        renderContext.SetState(blendStateType.Blend)

        ' Draw the transparent ruler
        renderContext.SetColorWireframe(Color.FromArgb(CInt(0.4 * 255), 255, 255, 255))
        renderContext.SetState(rasterizerStateType.CCW_PolygonFill_NoCullFace_NoPolygonOffset)

        ' Vertical Ruler
        renderContext.DrawQuad(New RectangleF(0, 0, rulerSize, CSng(Height - rulerSize)))

        ' Horizontal Ruler
        renderContext.DrawQuad(New RectangleF(rulerSize, CSng(Height - rulerSize), CSng(Width - rulerSize), rulerSize))

        renderContext.SetState(blendStateType.NoBlend)

        ' choose a string format with 2 decimal numbers
        Dim formatString As String = "{0:0.##}"

        Dim worldToScreen As Double = (Height - rulerSize) / (upperRightWorldCoord.Y - lowerLeftWorldCoord.Y)

        Dim stepWorldX As Double = 5, stepWorldY As Double = 5

        Dim worldHeight As Double = upperRightWorldCoord.Y - lowerLeftWorldCoord.Y
        Dim nlinesH As Double = (worldHeight / stepWorldY)

        Dim worldWidth As Double = upperRightWorldCoord.X - lowerLeftWorldCoord.X
        Dim nlinesW As Double = (worldWidth / stepWorldX)

        RefineStep(nlinesH, worldHeight, stepWorldY)
        RefineStep(nlinesW, worldWidth, stepWorldX)

        Dim stepWorld As Double = Math.Min(stepWorldX, stepWorldY)

        Dim stepScreen As Double = stepWorld * worldToScreen

        '''////////////////////////
        ' Vertical ruler
        '''////////////////////////

        ' First line Y world coordinate
        Dim startYWorld As Double = stepWorld * Math.Floor(lowerLeftWorldCoord.Y / stepWorld)

        Dim firstLineScreenPositionY As New Point2D(0, originScreen.Y + startYWorld * worldToScreen)
        Dim currentScreenY As Double = firstLineScreenPositionY.Y
        Dim shorterLineXPos As Double = (firstLineScreenPositionY.X + rulerSize) / 2

        ' draw a longer line each 5 lines. The origin must be a longer line.
        Dim countShortLinesY As Integer = CInt(Math.Round((currentScreenY - originScreen.Y) / stepScreen)) Mod 5

        ' Draw the ruler lines
        renderContext.SetLineSize(1)

        Dim left As Double

        Dim y As Double = startYWorld

        Dim myFont As Font = UtilityEx.GetFont(FontFamily, FontStyle, FontWeight, FontSize)

        While y < upperRightWorldCoord.Y
            If countShortLinesY Mod 5 = 0 Then
                left = firstLineScreenPositionY.X
            Else
                left = shorterLineXPos
            End If



            renderContext.SetColorWireframe(Color.Black)

            renderContext.DrawLine(New Point2D(left, currentScreenY), New Point2D(rulerSize, currentScreenY))

            DrawText(CInt(firstLineScreenPositionY.X), CInt(currentScreenY), String.Format(formatString, y), myFont, Color.Black, ContentAlignment.BottomLeft)

            countShortLinesY += 1
            y += stepWorld
            currentScreenY += stepScreen
        End While


        '''////////////////////////
        ' Horizontal ruler
        '''////////////////////////

        ' First line X world coordinate
        Dim startXWorld As Double = stepWorld * Math.Ceiling(lowerLeftWorldCoord.X / stepWorld)

        Dim firstLineScreenPositionX As New Point2D(originScreen.X + startXWorld * worldToScreen, 0)
        Dim currentScreenX As Double = firstLineScreenPositionX.X
        Dim shorterLineYPos As Double = (firstLineScreenPositionX.Y + rulerSize) / 2

        Dim countShortLinesX As Integer = CInt(Math.Round((currentScreenX - originScreen.X) / stepScreen)) Mod 5

        Dim top As Double
        Dim x As Double = startXWorld
        While x < upperRightWorldCoord.X
            If countShortLinesX Mod 5 = 0 Then
                top = firstLineScreenPositionX.Y
            Else
                top = shorterLineYPos
            End If

            renderContext.SetColorWireframe(Color.Black)

            renderContext.DrawLine(New Point2D(currentScreenX, Height - top), New Point2D(currentScreenX, Height - rulerSize))

            DrawText(CInt(currentScreenX), CInt(Height - rulerSize - firstLineScreenPositionX.Y), String.Format(formatString, x), myFont, Color.Black, ContentAlignment.BottomLeft)

            countShortLinesX += 1
            x += stepWorld
            currentScreenX += stepScreen
        End While


        ' Draw a red line in correspondence with the mouse position

        renderContext.SetColorWireframe(Color.Red)

        If mousePos.Y > rulerSize Then
            renderContext.DrawLine(New Point3D(0, Height - mousePos.Y, 0), New Point3D(rulerSize, Height - mousePos.Y, 0))
        End If

        If mousePos.X > rulerSize Then
            renderContext.DrawLine(New Point3D(mousePos.X, Height, 0), New Point3D(mousePos.X, Height - rulerSize, 0))
        End If

        ' Draw the logo image at the bottom right corner         
        Dim logo As new Bitmap("../../../../../../dataset/Assets/Pictures/Logo.png")
        DrawImage(Size.Width - logo.Width - 20, 20, logo)

        ' call the base function
        MyBase.DrawOverlay(myParams)
    End Sub

    Private Shared Sub RefineStep(nlines As Double, worldLength As Double, ByRef stepWorld As Double)
        ' Refine the step if there are too many ruler lines, or too few
        If nlines > 20 Then
            Do
                stepWorld *= 2
                nlines = (worldLength / stepWorld)
            Loop While nlines > 20
        ElseIf nlines < 10 Then
            Do
                stepWorld /= 2
                nlines = (worldLength / stepWorld)
            Loop While nlines < 10 AndAlso nlines > 0
        End If
    End Sub

    Dim mousePos as System.Drawing.Point 

    Protected Overrides Sub OnMouseMove(e As MouseEventArgs)
        MyBase.OnMouseMove(e)

        mousePos = RenderContextUtility.ConvertPoint(GetMousePosition(e))

        PaintBackBuffer()
        SwapBuffers()
    End Sub
End Class

